### DISCUSSION TITLE
Async Load CSS / Progressive Render

### DISCUSSION THREAD

**Discussion by lostra01** [withastro/roadmap#104] [Proposal]
### _Since the writing of this proposal, numerous issues have been outlined about the different options, thus I am rewriting this proposal to only include the information that is viable to implement._

-----------------------

### What's this proposal about

When working with Web, there are numerous way to get styling (CSS) into the HTML files themselves. There are two main ways to include external style sheets into the HTML files.

The first option being the regular `<link>` usage:

```html
<link rel="stylesheet" href="/path/to/my.css" />
```
the other option that can be used in HTML and the CSS files themselves is the `@import` option:
```html
<style>
@import url('/path/to/my.css');
</style>
```

### Using @import and why you shouldn't do it

When talking about performance, some nuances are often overlooked. One of them being using `@import` inside an HTML file, wrapped in style tags. While it seems convenient, the issue with this approach, is that it essentially removes the parallelisation of CSS file downloading. 

The reason for this, is because browsers implement an inert secondary parser, commonly referred to as Preload Scanner. The primary parser constructs DOM, CSSOM runs Javascript and is constantly stopping and starting as different parts of the site block it. 

The Preload Scanner can - should - safely jump ahead the primary parser, and scan the rest of the document to find additional references, then begin downloading them while the primary parser is still working on the previous items. The issue here is that @import is essentially hidden from the Preload Scanner if it's after any `<link>` or `<script>` tag. This causes any @imported CSS not to be downloaded in parallel, instead they will block the primary parser.
```html
<link rel="stylesheet" href="global.css" />

<style>
  @import url('/path/to/home.css');
</style>
```

To avoid any unnecessary bottlenecks, **do not use @import**. 

### Style tags in Astro

In any given `.astro` file, not one, but multiple `<style>` tags can be defined. What Astro does, is that it picks up any rules inside these style tags that are present on a page, and bundles them into a single CSS file with the naming of `{page}.{id}.css`. After it's done, Astro injects it into the `<head>` section of the site, using a regular Link tag, or it simply pushes the CSS rules into `<style>` tags into the head.

This is great, and not great at the same time. 

**Why this is good**

- All of our CSS is bundled into a single file
- If the end-user has good download speed the file will be delivered fast
- There will be no FOUC (Flash Of Unstyled Content) on the page since we do not start render until the CSS arrives
- There are no @imports only `<link>` or `<style>`

**The downsides**

- All of our CSS is bundled into a single file
- If the end-user has bad download speeds the file will be delivered over a long period of time
- We do not start render until the CSS arrives, thus on slow connections we might serve an empty pages for seconds, up to a minute
- We make no use of the available parallelism

### Solution

The first and most obvious solution is to utilize the parallelism of web browsers and split the CSS into multiple smaller ones - per Astro Component. This way instead of downloading one large CSS file in ~5 seconds, 5 ones could be downloaded in parallel completing the import in ~1 second. 

The only case when this option would yield worse performance, if the end-user's latency to the server is incredibly large, speaking over 100-300ms. In those cases, a single larger CSS file will outperform parallelism.

For a long time now, browsers support the feature to move `<link>` tags out of the header of the page. There are upsides and downsides to this as well. The link tags also have a set of `rel=` attributes they can use if they are in the body instead of the header. Luckily `rel="stylesheet"` is one of them. 

When a `<link>` tag is moved from the head to the body, the render of the document starts after the header components are loaded, and the links defined in the body will continue to download and be parsed/executed meanwhile. 

Theoritically, instead of linking in the header, like:

```html
...
<head>
    <link rel="stylesheet" href="/assets/global.css" />
    <link rel="stylesheet" href="/assets/header.css" />
    <link rel="stylesheet" href="/assets/navigation.css" />
    <link rel="stylesheet" href="/assets/content.css" />
    <link rel="stylesheet" href="/assets/footer.css" />
</head>
<body>

    <Header>
        <Navigation />
    </Header>
    
    <Content />
    
    <Footer />
    
</body>
...
``` 
it is possible - and in some cases preferred - to add the links to the body, just before the actual component.
```html
...
<head>
    <link rel="stylesheet" href="/assets/global.css" />
</head>
<body>

    <link rel="stylesheet" href="/assets/header.css" />
    <Header>
    
        <link rel="stylesheet" href="/assets/navigation.css" />
        <Navigation />
    </Header>
    
    <link rel="stylesheet" href="/assets/content.css" />
    <Content />
    
    <link rel="stylesheet" href="/assets/footer.css" />
    <Footer />
    
</body>
...
``` 
What this does, is will let the browser load the `global.css` file, then start the rendering of the page. While the `global.css` file is still loading, the Preload Scanner will go ahead, and in order or appearance start downloading the rest of the CSS files, in parallel. So by the time rendering reaches the next component, the CSS is already loaded and the content is properly displayed.

**Note:** _In an non-optimal case - high latency / slow network - it is possible that the rest of the CSS is not downloaded fast enough, and heavy FOUC happens. This is obviously undesired behaviour. We can get around this, as outlined in the next section._

### **Proposal**

Since Astro components handle multiple `<style>` tags, we could give the developers the ability to manually opt-in to using the new feature, using the `async` directive on a `<style>` tag, such as:

```html
<style async>
    .myComponent{
        width: 100vw;
        background-color: gray;
    }
</style>
```

This would cause the component's style to be bundled in a separate CSS file, and to be injected into the body, right before the component injection happens, turning this `.astro` example:

```astro
---
const {Welcome} = Astro.props
---

<div class="welcome"> {Welcome} </div>

<style async>
    .welcome{
        font-size: 24px;
        color: blue;
    }
</style>
```

when built, into the following HTML snippet:

```html
...
<head>
    <link rel="stylesheet" href="/assets/global.css" />
</head>
<body>

    <link rel="stylesheet" href="/assets/welcome.css" />
    <div class="welcome"> Welcome to my page! </div>
    
</body>
...
```

Adding this optional directive gives the developers full control on what stylesheets can be loaded in the body, and risk possible FOUC, and what should be kept in the head. Since Astro supports multiple `<style>` tags, it is possible to add two of them into a component, one being a regular style, that handles the sizing of the component and it's parts, with the other being an async style, that contains elements that are not necessary or non critical from the rendering aspect of the page.

```astro
---
const {Welcome} = Astro.props
---

<div class="welcome"> {Welcome} </div>

<style>
    .welcome{
        width: 100%;
        text-align: center;
        font-size: 24px;
    }
</style>
<style async>
    .welcome{
        text-decoration: underline;
        color: blue;
    }
</style>
```

```html
...
<head>
    <link rel="stylesheet" href="/assets/global.css" /> <!-- Contains width, text-align and font-size for .welcome -->
</head>
<body>

    <link rel="stylesheet" href="/assets/welcome.css" /> <!-- Contains color and text-decoration for .welcome -->
    <Welcome />
    
</body>
...
```

This way the default behavior of Astro stays the same, so there will be no must-learn new feature, and developers can manually opt-in to using async styles inside their component. 

--------------------------------
### Resources

https://csswizardry.com/2018/11/css-and-network-performance/
# https://github.com/withastro/rfcs/blob/main/proposals/0008-style-script-behavior.md

**Comment by aFuzzyBear** [comment]
I would like to also tie this into the wider discussion surrounding the following RFC on formalising the `<style>` behaviour,
https://github.com/withastro/rfcs/blob/main/proposals/0008-style-script-behavior.md

**Reply by lostra01** [reply]
Thanks for the link! I forgot to include that as well. I'll be also inserting here some valid points we discussed about this so others can join in as well:

The RFC on `<style>` and `<script>` behaviors includes a `global` directive for the style tags, which essentially removes scoping. 
Given the way how asyncing the styles could work means that `global` and `async` directives would essentially knock each other out, so it is important to make sure those two cannot be used together.

Another aspect that mentioned if there's any differentiation between using the `<style>` tags in Layouts, Pages or Components. As far as I know at the moment Astro makes no difference based on where the style is coming from, and they are all bundled into the same file and then included in the `<head>`. In my opinion, I would keep this behavior.

According to the findings in the original proposals and the further discussion about them, it comes down to one plausible solution that would add value to Astro instead of introducing new issues or placing burden on the developers:

- To avoid changing the expected / used-to styling in Astro, developers would have to manually opt-in to use the feature with a manual directive. This way if someone has no information about said feature, and does not want to learn about said feature, they can keep using Astro as it is, and there will be no behavioral changes.

- `global` and `async` directives on `<style>` tags could not be used at the same time.

- Inject `async` styles with their generated CSS into the `<body>` tag, right before the content injection of the component.

- Using multiple `<style>` tags inside components is already supported by Astro, which means it is possible to have some styling of the component be present in the `<head>` tag and less important ones can be deferred. Think of a column display. Users can specify two columns, each with a width of 50% in a regular `<style>` tag, while putting font colors in an `async` tag. This way the worst FOUC that can happen is the one the developers explicitly allow, like font colors, text decoration, etc.

- Using a single `<style>` tag inside a component with `async` directive on it lets developers completely defer the load of the component's CSS, which can be a larger performance gain when used on components that have no chance of FOUC like footers, etc.

**Comment by FredKSchott** [comment]
## Notes from RFC Call:

- `async` vs. `inline` (naming very much TBD!)
- how does this compare to a site-level optimization pass? For example, paired with: https://github.com/GoogleChromeLabs/critters
  - This could be available if you want it, but then ignore it if you don't
- more exploration needed, but general excitement for more control over CSS optimization

**Comment by peckomingo** [comment]
I have a Layout `BaseLayout.astro` and i want it to have a script inside the <head> which is available on all pages, so i use 
```
<script async hoist>
    import '../../assets/js/test.ts'
</script>
```

It works fine, but the async attribute gets removed in the output after running npm run build.

I think it is the similar or same issue as described here for the CSS case.

**Comment by florian-lefebvre** [comment]
Closing because this is outdated
